### PR TITLE
TIP-451: Return standard format in product & variant group processors

### PR DESCRIPTION
**Pull request by skeleton**
- [x] return standard format in product processor
- [x] return standard format in quick product processor
- [x] return standard format in variant group processor
- [x] manage media for product & variant group writers

Bonus:
- [x] media are now exported on a quick export

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       |
| Added Behats                      |
| Changelog updated                 |
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |
| Migration script                  |
| Tech Doc                          |

### CODE REVIEW

File: src/Pim/Component/Connector/Writer/File/BulkFileExporter.php [hunk 1]
```diff
@@ -17,33 +19,65 @@ class BulkFileExporter
...
+                        'scope'  => $value->getScope()
+                    ],
+                    [
+                        'identifier' => $identifier,
+                        'code'       => $value->getAttribute()->getCode()
+                    ]
+                );
+
+                $this->doCopy([
+                    'to'      => $exportPath . $medium->getOriginalFilename(),
```
Comment (skeleton) [inline src/Pim/Component/Connector/Writer/File/BulkFileExporter.php]: inverse to & from for juju

**Comment by skeleton** [comment]
change media/medium

File: features/export/configure_export_products_with_media.feature [hunk 2]
```diff
@@ -41,7 +42,7 @@ Feature: Configure export of products media
     And I launch the export job
     And I wait for the "csv_footwear_product_export" job to finish
     Then I should not see the text "Download generated archive"
-    And export directory of "csv_footwear_product_export" should contain the following media:
+    And export directory of "csv_footwear_product_export" should contain the following file:
       | product_export.csv                       |
```
Comment (pierallard) [inline features/export/configure_export_products_with_media.feature]: extra blank

File: src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml [hunk 3]
```diff
@@ -26,6 +26,7 @@ parameters:
     pim_connector.array_converter.standard_to_flat.group.class:                                   Pim\Component\Connector\ArrayConverter\StandardToFlat\Group
     pim_connector.array_converter.standard_to_flat.group_type.class:                              Pim\Component\Connector\ArrayConverter\StandardToFlat\GroupType
     pim_connector.array_converter.standard_to_flat.product.class:                                 Pim\Component\Connector\ArrayConverter\StandardToFlat\Product
+    pim_connector.array_converter.standard_to_flat.product_localized.class:                       Pim\Component\Connector\ArrayConverter\StandardToFlat\ProductLocalized
```
Comment (pierallard) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml]: I had rather written "localized_product"

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    /** @var ChannelRepositoryInterface */
+    protected $channelRepository;
+
+    /** @var AttributeRepositoryInterface */
+    protected $attributeRepository;
+
+    /** @var ProductBuilderInterface */
+    protected $productBuilder;
+
+    /** @var  ObjectDetacherInterface */
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: double space :eyeglasses:

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        $productStandard = $this->normalizer->normalize($product, 'json', $normalizerContext);
+        $productStandard = $this->filterProperties(
+            $productStandard,
+            $normalizerContext['localeCodes'],
+            $normalizerContext['scopeCode'],
+            $normalizerContext['selected_properties']
+        );
+
+        $parameters = $this->stepExecution->getJobParameters();
+        if ($parameters->has('with_media') && $parameters->get('with_media')) {
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: Just a question (not sure). Should we write `true === $parameters->get('with_media')` or not ?

Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: `with_media` has been defined as a boolean, so we should not add `true ===` (we never do this in the PIM)

File: src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml (hunk 3 above)
Comment (skeleton) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml]: I keep the same naming as https://github.com/akeneo/pim-community-dev/blob/master/src/Pim/Bundle/ConnectorBundle/Resources/config/array_converters.yml#L5

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+    /**
+     * @param array      $values
+     * @param array      $localeCodes
+     * @param string     $channelCode
+     * @param array|null $selectedProperties
+     *
+     * @return array
+     */
+    protected function filterValues(
```
Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: missing description

File: src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php [hunk 7]
```diff
@@ -39,17 +42,41 @@ function it_normalizes_groups(
...
+                'metric' => ['data' => 12.5000, 'locale' => null, 'scope' => null],
+                'prices' => ['data' => 12.5, 'locale' => null, 'scope' => null],
+                'date'   => ['data' => '2015-01-31', 'locale' => null, 'scope' => null],
+            ]
+        ];
+
+        $valuesLocalized = [
+            'number' => ['data' => '12,5000', 'locale' => null, 'scope' => null],
+            'metric' => ['data' => '12,5000', 'locale' => null, 'scope' => null],
+            'prices' => ['data' => '12,5', 'locale' => null, 'scope' => null],
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php]: Number and Metrics are displayed with 4 numbers, price just 1 ?

File: src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php [hunk 8]
```diff
@@ -68,25 +95,26 @@ function () use (&$productsCount) {
...
-                'model_type'        => 'variant_group',
-                'created'           => 'normalized_oldest_log',
-                'updated'           => 'normalized_newest_log',
+        $this->normalize($tshirt, 'internal_api', $options)->shouldReturn(
+            [
+                'code'     => 'my_variant',
+                'axis'     => ['color', 'size'],
+                'type'     => 'variant',
+                'values'   => $valuesLocalized,
+                'products' => [42],
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php]: I don't think it's the scope of the card, but it's not products, it's productIds. To see with @jjanvier

File: src/Pim/Bundle/ConnectorBundle/Resources/config/processors.yml [hunk 9]
```diff
@@ -188,14 +189,23 @@ services:
...
 
+    pim_connector.processor.normalization.product:
+        class: '%pim_connector.processor.normalization.product.class%'
+        arguments:
+            - '@pim_serializer.normalizer.product'
+            - '@pim_catalog.repository.channel'
+            - '@pim_catalog.repository.attribute'
+            - '@pim_catalog.builder.product'
+            - '@akeneo_storage_utils.doctrine.object_detacher'
+            - '@pim_connector.writer.file.product.bulk_file_exporter'
```
Comment (jjanvier) [inline src/Pim/Bundle/ConnectorBundle/Resources/config/processors.yml]: ok, we'll move the file_exporters in a separate PR

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+
+        foreach ($mediaAttributeTypes as $attributeCode => $attributeType) {
+            foreach ($item['values'][$attributeCode] as $index => $value) {
+                if (null !== $value['data']) {
+                    $exportDirectory = $this->fileExporterPath->generate($value, [
+                        'identifier' => $identifier,
+                        'code'       => $attributeCode,
+                    ]);
+
+                    $finder = new Finder();
```
Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: maybe explain in the method descr why you need to do this ?

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductValueNormalizer.php [hunk 11]
```diff
@@ -58,11 +52,9 @@ public function normalize($entity, $format = null, array $context = [])
             $data = $this->serializer->normalize($entity->getData(), $format, $context);
         }
 
-        $type = $entity->getAttribute()->getAttributeType();
-
-        $localizer = $this->localizerRegistry->getLocalizer($type);
-        if (null !== $localizer) {
-            $data = $localizer->localize($data, $context);
+        if (AttributeTypes::NUMBER === $attribute->getAttributeType() && null !== $data && is_numeric($data)) {
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductValueNormalizer.php]: Should we add a specific normalizer for NUMBER attribute type ?

File: src/Pim/Component/Connector/Writer/File/BulkFileExporter.php [hunk 12]
```diff
@@ -17,33 +19,65 @@ class BulkFileExporter
...
-    public function exportAll(array $items, $target)
+    public function exportAll(ArrayCollection $items, $target, $identifier)
     {
-        foreach ($items as $media) {
-            foreach ($media as $medium) {
-                $this->doCopy($medium, $target);
+        foreach ($items as $value) {
+            if (!$value instanceof ProductValueInterface) {
+                throw new \InvalidArgumentException(
+                    'Value is not an instance of Pim\Component\Catalog\Model\ProductValueInterface.'
```
Comment (solivier) [inline src/Pim/Component/Connector/Writer/File/BulkFileExporter.php]: sprintf('Value is not an instance of %s.', ProductValueInterface::class) ?

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product.php [hunk 13]
```diff
@@ -38,19 +38,22 @@ protected function convertProperty($property, $data, array $convertedItem, array
                 $convertedItem[$property] = implode(',', $data);
                 break;
             case 'enabled':
+                $convertedItem[$property] = false === $data || null === $data ? '0' : '1';
```
Comment (pierallard) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product.php]: (false === $data || null === $data)
Maybe you can replace it by `!$data`.

**Comment by solivier** [comment]
gtm

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+        $propertiesToExport = [];
+        foreach ($product as $codeProperty => $property) {
+            if ('values' === $codeProperty) {
+                $propertiesToExport['values'] = $this->filterValues(
+                    $property,
+                    $localeCodes,
+                    $channelCode,
+                    $selectedProperties
+                );
+            } elseif (null === $selectedProperties || in_array($codeProperty, $selectedProperties)) {
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: imho, we shound never receive `null` to define we don't select anything
it should be `[]`

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+            $this->importMedia($product, $directory);
+        }
+
+        $this->detacher->detach($product);
+
+        return $productStandard;
+    }
+
+    /**
+     * Filter values to keep only selected values with scope & locales defined by context
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: explain that null when no properties has been selected, which means we export everything

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+use Pim\Component\Connector\ArrayConverter\ArrayConverterInterface;
+
+/**
+ * Convert standard format to flat format for product with localized values.
+ *
+ * @author    Marie Bochu <marie.bochu@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ProductLocalized implements ArrayConverterInterface
```
Comment (pierallard) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php]: I don't understand the name of this class :/

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php (hunk 15 above)
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: properties here, not values

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+            } elseif (null === $selectedProperties || in_array($codeProperty, $selectedProperties)) {
+                $propertiesToExport[$codeProperty] = $property;
+            }
+        }
+
+        return $propertiesToExport;
+    }
+
+    /**
+     * @param array      $values
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: Filter values to keep only selected values with scope & locales defined by context

File: src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php (hunk 7 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php]: metric & number have 4 decimals.
price only 2 #consistency

this spec does not represent the real case, I'll fix it

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+        return $valuesToExport;
+    }
+
+    /**
+     * Import media on the local filesystem
+     *
+     * @param ProductInterface $product
+     * @param string           $directory
+     */
+    protected function importMedia(ProductInterface $product, $directory)
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: fecthMediaS? to avoid confusion

we are in a processor, you call a function `import` and this `import` function uses an `exporter` ^^

btw, we'll move/rename these exporters later

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+
+    /**
+     * @throws \InvalidArgumentException
+     *
+     * @return array
+     */
+    protected function getNormalizerContext()
+    {
+        $jobParameters = $this->stepExecution->getJobParameters();
+        $mainContext = $jobParameters->get('mainContext');
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: don't understand why we have a main context here
why not directly have locale and channel?

not to do here :)

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+        $columns = $jobParameters->get('selected_properties');
+
+        if (!isset($mainContext['scope'])) {
+            throw new \InvalidArgumentException('No channel found');
+        }
+
+        if (isset($columns) && 0 !== count($columns)) {
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: null !== $columns

because $columns is set 5 lines above

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+    /** @var BufferFactory */
+    protected $bufferFactory;
+
+    /** @var AttributeRepositoryInterface */
+    protected $attributeRepository;
+
+    /** @var FileExporterPathGeneratorInterface */
+    protected $fileExporterPath;
+
+    /** @var array */
```
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: string[]

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+    protected $filePathResolverOptions;
+
+    /** @var Filesystem */
+    protected $localFs;
+
+    /** @var array */
+    protected $writtenFiles = [];
+
+    /** @var FlatItemBuffer */
+    protected $flatRowBuffer = null;
```
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: Drop null

File: src/Pim/Component/Catalog/Normalizer/Structured/MetricNormalizer.php [hunk 23]
```diff
@@ -22,8 +24,14 @@ class MetricNormalizer implements NormalizerInterface
      */
     public function normalize($object, $format = null, array $context = [])
     {
+        $data = $object->getData();
+        if (null !== $data && is_numeric($data) && isset($context['decimals_allowed'])) {
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/MetricNormalizer.php]: add a TODO and explains
 * to be ISO mongo/orm
 * we decided to do it in the standard format right now but that it could/should be changed

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductPriceNormalizer.php [hunk 24]
```diff
@@ -22,8 +24,14 @@ class ProductPriceNormalizer implements NormalizerInterface
      */
     public function normalize($object, $format = null, array $context = [])
     {
+        $data = $object->getData();
+        if (null !== $data && is_numeric($data) && isset($context['decimals_allowed'])) {
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductPriceNormalizer.php]: same here

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+    ) {
+        $this->arrayConverter = $arrayConverter;
+        $this->bufferFactory = $bufferFactory;
+        $this->flusher = $flusher;
+        $this->attributeRepository = $attributeRepository;
+        $this->mediaAttributeTypes = $mediaAttributeTypes;
+        $this->fileExporterPath = $fileExporterPath;
+
+        $this->filePathResolver = new OptionsResolver();
+        $this->filePathResolver->setRequired('parameters');
```
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: I already heard to put all the option resolver conf into a dedicated function... Not sure if it's the "akeneo" process.

File: src/Pim/Component/Catalog/Normalizer/Structured/ProductValueNormalizer.php (hunk 11 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Normalizer/Structured/ProductValueNormalizer.php]: same

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product.php (hunk 13 above)
Comment (jjanvier) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product.php]: sure? shouldn't it be '' in case $data is null?

File: src/Pim/Component/Connector/Writer/File/BulkFileExporter.php (hunk 12 above)
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/BulkFileExporter.php]: use "get class" function please

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product/ValueConverter/BooleanConverter.php [hunk 26]
```diff
@@ -44,7 +42,7 @@ public function convert($attributeCode, $data)
                 $value['scope']
             );
 
-            $convertedItem[$flatName] = (true === $value['data']) ? '1' : '0';
+            $convertedItem[$flatName] = false === $value['data'] || null === $value['data'] ? '0' : '1';
```
Comment (jjanvier) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product/ValueConverter/BooleanConverter.php]: sure? shouldn't it be '' in case $data is null?

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+        $this->localizer = $localizer;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function convert(array $productStandard, array $options = [])
+    {
+        $productStandard['values'] = $this->localizer->convertToLocalizedFormats($productStandard['values'], $options);
+        $productFlat = $this->converter->convert($productStandard, $options);
```
Comment (jjanvier) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php]: can't have violations here?
in case that happens, you should return a DataArrayConversionException

File: src/Pim/Component/Connector/Writer/File/BulkFileExporter.php [hunk 28]
```diff
@@ -70,71 +102,35 @@ public function getErrors()
...
-                $this->addCopiedMedium($medium, $target);
-            } catch (FileTransferException $e) {
-                $this->addError($medium, 'The media has not been found or is not currently available');
-            } catch (\LogicException $e) {
-                $this->addError($medium, sprintf('The media has not been copied. %s', $e->getMessage()));
-            }
+        try {
+            $this->fileExporter->export($media['from'], $target, $media['storage']);
+        } catch (FileTransferException $e) {
+            $this->addError($media, 'The media has not been found or is not currently available');
```
Comment (pierallard) [inline src/Pim/Component/Connector/Writer/File/BulkFileExporter.php]: I think you should add the $e->getMessage() at the end of this sentence (for better error sourcing)

File: src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+        $parameters = $this->stepExecution->getJobParameters();
+        $structure =  $parameters->get('filters')['structure'];
+        $channel = $this->channelRepository->findOneByIdentifier($structure['scope']);
+        $this->productBuilder->addMissingProductValues($product, [$channel], $channel->getLocales()->toArray());
+
+        $attributesToFilter = $this->getAttributesToFilter($parameters);
+        $productStandard = $this->normalizer->normalize($product, 'json');
+        $productStandard['values'] = $this->filterValues(
+            $productStandard['values'],
+            array_intersect($channel->getLocaleCodes(), $structure['locales']),
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php]: that means you always have the locales
even if all the locales are exported

that's it?

File: src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php [hunk 30]
```diff
@@ -0,0 +1,197 @@
...
+            $channel->getCode(),
+            $attributesToFilter
+        );
+
+        if ($parameters->has('with_media') && $parameters->get('with_media')) {
+            $directory = $this->getWorkingDirectory($parameters->get('filePath'));
+            $this->importMedia($product, $directory);
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php]: same remark about the naming that in quick export processor

File: src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php [hunk 31]
```diff
@@ -0,0 +1,197 @@
...
+                    }
+                );
+            }
+        }
+
+        return $valuesToExport;
+    }
+
+    /**
+     * Return a list of attributes to export
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php]: either explain it returns null when you have to export all the filters

OR

never return null
do another method `areAttributesToFilter` that returns a bool, and use to know if you have to filter things or not

I think you know what I prefer, but it's up to you ^^ :)

File: src/Pim/Component/Connector/spec/Writer/File/Csv/ProductWriterSpec.php [hunk 32]
```diff
@@ -179,4 +307,17 @@ function it_writes_the_csv_file(
...
+    function it_builds_the_path(StepExecution $stepExecution, JobParameters $jobParameters)
+    {
+        $options = ['date' => '2015-01-01'];
+        $this->setStepExecution($stepExecution);
+        $stepExecution->getJobParameters()->willReturn($jobParameters);
+        $jobParameters->has('mainContext')->willReturn(true);
+        $jobParameters->get('mainContext')->willReturn($options);
+        $jobParameters->get('filePath')->willReturn($this->directory . 'product_%date%.csv');
+
+
```
Comment (pierallard) [inline src/Pim/Component/Connector/spec/Writer/File/Csv/ProductWriterSpec.php]: double empty line

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php (hunk 22 above)
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: useless null

File: src/Pim/Component/Connector/spec/Writer/File/Xlsx/ProductWriterSpec.php [hunk 33]
```diff
@@ -4,205 +4,320 @@
...
+        $flatRowBuffer->write([$productFlat1, $productFlat2], ['withHeader' => true])->shouldBeCalled();
 
-        $stepExecution->addWarning(Argument::cetera())->shouldBeCalled();
         $this->initialize();
         $this->write($items);
 
         $this->getWrittenFiles()->shouldBeEqualTo([
-            '/tmp/export' => 'export'
-        ]);
+                                                      $productPathMedia1 . 'it\'s the filename.jpg' => 'files/jackets/media/it\'s the filename.jpg'
```
Comment (pierallard) [inline src/Pim/Component/Connector/spec/Writer/File/Xlsx/ProductWriterSpec.php]: wut?

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php [hunk 34]
```diff
@@ -0,0 +1,293 @@
...
+     * Return the identifier of the item (e.q sku or variant group code)
+     *
+     * @param array $item
+     *
+     * @return string
+     */
+    abstract protected function getIdentifier(array $item);
+
+    /**
+     * Stock media for archiver and change media path
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: Explain here with an example
And do the same in the processor please

File: src/Pim/Component/Connector/spec/Writer/File/Xlsx/VariantGroupWriterSpec.php [hunk 35]
```diff
@@ -4,142 +4,223 @@
...
+        $arrayConverter->convert($variantStandard1, [])->willReturn($variantFlat1);
+        $arrayConverter->convert($variantStandard2, [])->willReturn($variantFlat2);
+
+        $flatRowBuffer->write([$variantFlat1, $variantFlat2], ['withHeader' => true])->shouldBeCalled();
+
+        $this->initialize();
+        $this->write($items);
+
+        $this->getWrittenFiles()->shouldBeEqualTo([
+                                                      $variantPathMedia1 . 'it\'s the filename.jpg' => 'files/jackets/media/it\'s the filename.jpg'
```
Comment (pierallard) [inline src/Pim/Component/Connector/spec/Writer/File/Xlsx/VariantGroupWriterSpec.php]: !

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php [hunk 36]
```diff
@@ -0,0 +1,293 @@
...
+    {
+        $this->stepExecution = $stepExecution;
+    }
+
+    /**
+     * Get configuration for writer (type of export, delimiter, enclosure, etc)
+     *
+     * @return array
+     */
+    abstract protected function getConfigurationWriter();
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: getWriterConfiguration

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php [hunk 37]
```diff
@@ -0,0 +1,293 @@
...
+
+    /**
+     * Return the identifier of the item (e.q sku or variant group code)
+     *
+     * @param array $item
+     *
+     * @return string
+     */
+    abstract protected function getIdentifier(array $item);
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: getItemIdentifier

File: src/Pim/Component/Connector/Writer/File/MediaExporterPathGenerator.php [hunk 38]
```diff
@@ -2,9 +2,6 @@
 
 namespace Pim\Component\Connector\Writer\File;
 
-use Doctrine\Common\Util\ClassUtils;
-use Pim\Component\Catalog\Model\ProductValueInterface;
-
 /**
  * Generate the path of medias to export.
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Writer/File/MediaExporterPathGenerator.php]: it does not generate the path of the media anymore
to clean in another story

**Comment by jjanvier** [comment]
Bravo :+1: . A huge step forwards :)
GTM once fixed :)

File: src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php (hunk 8 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/spec/Normalizer/GroupNormalizerSpec.php]: in fact, we never have a reference to products linked to a variant group in standard format

**Comment by pierallard** [comment]
GTM with some comments :) It's fucking good to have this done.

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product.php (hunk 13 above)
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/Product.php]: @pierallard nope :) I prefer to be explicit 
@jjanvier yes I'm sure :( if a boolean attribute is in a family, it can be null, but in exports we put '0'

File: src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php (hunk 19 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Connector/Processor/QuickExport/ProductProcessor.php]: I agree to change it (I just copy this code), but yeah, not in this PR :)

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php (hunk 27 above)
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php]: there is no violation when you localize data

File: src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php (hunk 16 above)
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/StandardToFlat/ProductLocalized.php]: what do you don't understand ?

File: src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php (hunk 10 above)
Comment (skeleton) [inline src/Pim/Component/Connector/Writer/File/AbstractItemMediaWriter.php]: done

File: src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php (hunk 31 above)
Comment (skeleton) [inline src/Pim/Component/Connector/Processor/Normalization/ProductProcessor.php]: done ;)

**Comment by nidup** [comment]
Thx @skeleton best PR à un company. Ever.
